osu.Game/Screens/Play/GameplayClockContainer.cs

```diff
@@ -1,300 +1,126 @@
         {
-            if (!adjustableClock.IsRunning)
+            if (!AdjustableSource.IsRunning)
             {
                 // Seeking the decoupled clock to its current time ensures that its source clock will be seeked to the same time
-                // This accounts for the audio clock source potentially taking time to enter a completely stopped state
+                // This accounts for the clock source potentially taking time to enter a completely stopped state
                 Seek(GameplayClock.CurrentTime);
 
-                adjustableClock.Start();
+                AdjustableSource.Start();
             }
 
             IsPaused.Value = false;
         }
 
-        /// <summary>
-        /// Skip forward to the next valid skip point.
-        /// </summary>
-        public void Skip()
-        {
-            if (GameplayClock.CurrentTime > gameplayStartTime - MINIMUM_SKIP_TIME)
-                return;
-
-            double skipTarget = gameplayStartTime - MINIMUM_SKIP_TIME;
-
-            if (GameplayClock.CurrentTime < 0 && skipTarget > 6000)
-                // double skip exception for storyboards with very long intros
-                skipTarget = 0;
-
-            Seek(skipTarget);
-        }
-
         /// <summary>
         /// Seek to a specific time in gameplay.
-        /// <remarks>
-        /// Adjusts for any offsets which have been applied (so the seek may not be the expected point in time on the underlying audio track).
-        /// </remarks>
         /// </summary>
         /// <param name="time">The destination time to seek to.</param>
-        public void Seek(double time)
-        {
-            // remove the offset component here because most of the time we want the seek to be aligned to gameplay, not the audio track.
-            // we may want to consider reversing the application of offsets in the future as it may feel more correct.
-            adjustableClock.Seek(time - totalOffset);
+        public virtual void Seek(double time) => AdjustableSource.Seek(time);
 
-            // manually process frame to ensure GameplayClock is correctly updated after a seek.
-            userOffsetClock.ProcessFrame();
-        }
-
-        public void Stop()
-        {
-            IsPaused.Value = true;
-        }
+        /// <summary>
+        /// Stops gameplay.
+        /// </summary>
+        public virtual void Stop() => IsPaused.Value = true;
 
         /// <summary>
-        /// Changes the backing clock to avoid using the originally provided track.
+        /// Restarts gameplay.
         /// </summary>
-        public void StopUsingBeatmapClock()
+        public virtual void Restart()
         {
-            removeSourceClockAdjustments();
+            AdjustableSource.Seek(0);
+            AdjustableSource.Stop();
 
-            track = new TrackVirtual(track.Length);
-            adjustableClock.ChangeSource(track);
+            if (!IsPaused.Value)
+                Start();
         }
```

Comment:
Right, I forgot to check this.

- I looked back through blames and found no good reasoning for why this is in a `Task.Run()`. I think it was probably just performance, but it's a few milliseconds at worst, and it's not like everything here is async'd.
- ~~The source change is done in the ctor as above.~~ (see above discussion)
- Rate update is not scheduled because this is completely running on the update thread.

Reply:
Removing this schedule broke the reset procedure [as I described on discord](https://discord.com/channels/188630481301012481/188630652340404224/834839320157224961). I am not yet 100% unsure why, and am looking to submit a fix for this, but am leaving this comment in case I fail to do so in the time I have left.